### DISCUSSION TITLE
Respect GDPR for broker users

### DISCUSSION THREAD

**Discussion by cgeorgilakis** [keycloak/keycloak#30193] [Ideas]
Keycloak already supports GDPR during new User registration asking for accepting terms and conditions.  However, when a new broker user attempts to register/sign in using the first-broker-login flow in Keycloak, personal data such as username, name, or email (if provided by the Identity Provider) is stored before the user explicitly consents to the terms and conditions. This practice violates the GDPR requirements in Europe, which mandate that explicit consent must be obtained from the user before any personal data is collected, stored, or processed.

The relevant clause from the GDPR regarding consent is Article 6, which outlines the lawfulness of processing. Specifically, Article 6(1)(a) states:

> Processing shall be lawful only if and to the extent that at least one of the following applies: (a) the data subject has given consent to the processing of his or her personal data for one or more specific purposes.

Furthermore, Article 7 details the conditions for consent, emphasising that consent must be freely given, specific, informed, and unambiguous.

Additionally, Recital 32 of the GDPR provides further clarification on consent, stating:

> Consent should be given by a clear affirmative act establishing a freely given, specific, informed, and unambiguous indication of the data subject's agreement to the processing of personal data relating to him or her...

These provisions collectively underscore the requirement for obtaining explicit consent before collecting or processing personal data.

A [github issue](https://github.com/keycloak/keycloak/issues/28714) has already existed for this. Could you review the important [PR](https://github.com/keycloak/keycloak/pull/29975) we have opened in order to make Keycloak follow European law?

**Comment by hmlnarik** [comment]
Would it be possible to leverage transient users feature (#26637) to achieve GDPR compliance? Keycloak does not store any data about a transient user in its database, the user data are discarded when the user is logged out (together with deletion of the user session).

**Reply by cgeorgilakis** [reply]
> Would it be possible to leverage transient users feature (#26637) to achieve GDPR compliance? Keycloak does not store any data about a transient user in its database, the user data are discarded when the user is logged out (together with deletion of the user session).

No.
We want this user after accepting terms and conditions to be saved in Keycloak database.
This will mean that he will be able to be added in a group, link with another Identity Provider, log in with another client etc. For sure we do not want he registers/signs inagain every time. He will be in Keycloak database.

**Comment by cgeorgilakis** [comment]
During **Keycloak-DevDay Pre-Conf Event**, we have discussed with some participants and it was common sense that Keycloak must support asking federated users to accept `Terms & Conditions` before saving user to database. This is needed not only for respecting **GDPR** in Europe but also this is needed in other use cases such as **HIPAA** compliance that refers to the process by which healthcare organizations and their business associates adhere to the Health Insurance Portability and Accountability Act (HIPAA), a US federal law, to protect the privacy and security of Protected Health Information (PHI) (@mihirvadalia ).

We need to decide the way, Keycloak will be able to ask for `Terms & Conditions`. We think for two possible ways :
-	As in [my PR](https://github.com/keycloak/keycloak/pull/29975), user will be able to accept `Terms & Conditions` together with any user attribute changes (Review Profile step). As our team has investigated, for `Update Profile on First Login ` with value `on ` together with `Accept Terms and Conditions`, a user can change readonly user attributes and pass new values to Keycloak even he is not allowed. So, [my PR](https://github.com/keycloak/keycloak/pull/29975) is in draft mode, trying to make impossiblechanging Identity Provider values for `Update Profile on First Login ` with value `on ` . If we agree to proceed with this way, appropriate changes is need in my PR.
-	Second way is by separate `terms.ftl ` before user is saved in database. Such solution describes in @BlackVoid [comment](https://github.com/keycloak/keycloak/issues/28714#issuecomment-2703274296) in the related [issue](https://github.com/keycloak/keycloak/issues/28714).  @BlackVoid Could you make it PR if Keycloak team agree to proceed with this way?

Keycloak team and Keycloak users what do you believe is the best way to proceed with being possible to accept before user is saved for federated users? @thomasdarimont @edwint88

**Reply by BlackVoid** [reply]
I'd be able to make a PR if that's the way forward, but I'd probably need guidance on the reject part of the solution, since there are authentication flows that my solution is incompatible with, due to some variables not being available (from what I can tell).

**Reply by cgeorgilakis** [reply]
For me, rejecting is meaning user will not created in database and flow will be finished.

**Comment by stianst** [comment]
Ability to ask user to fill terms and conditions before creating the user would be a good thing; I don't think accepting terms and conditions should be an authentication flow thing, and rather just a general action. One challenge is that required actions currently can only run after authentication has completed. What I could imagine is if we had some way of running actions on different stages, then that could be used for a general terms and conditions action. If user is present it can store it in the user, if not it can add it to the session, then the action triggers again after the user has authenticated and stores if need.

I would also think that there are cases where you would want a user to accept terms and conditions prior to filling in the user information and creating the user.

**Reply by cgeorgilakis** [reply]
> Ability to ask user to fill terms and conditions before creating the user would be a good thing; I don't think accepting terms and conditions should be an authentication flow thing, and rather just a general action. One challenge is that required actions currently can only run after authentication has completed. What I could imagine is if we had some way of running actions on different stages, then that could be used for a general terms and conditions action. If user is present it can store it in the user, if not it can add it to the session, then the action triggers again after the user has authenticated and stores if need.
> 
> I would also think that there are cases where you would want a user to accept terms and conditions prior to filling in the user information and creating the user.

If someone could implement the change in required actions (mainly for terms and conditions), this could be a solution.

**Comment by stianst** [comment]
Since transient users was mentioned here (and I know it's not a solution to this problem) I'll mention that is most likely going to be removed and replaced with https://github.com/keycloak/keycloak/issues/38108
